torch/fx/node.py

```diff
@@ -890,35 +891,41 @@ def __setattr__(self, name: str, value: Any) -> None:
 
 
 @compatibility(is_backward_compatible=True)
-def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
+def map_arg(a: ArgumentT, fn: Callable[[Node], Argument]) -> ArgumentT:
     """
-    Apply fn to each Node appearing arg. arg may be a list, tuple, slice, or dict with string keys.
+    Apply fn recursively to each Node appearing in arg.
+
+    arg may be a list, tuple, slice, or dict: the return value will have
+    the same type and structure.
     """
     assert callable(fn), "torch.fx.map_arg(a, fn): fn must be a callable"
     return map_aggregate(a, lambda x: fn(x) if isinstance(x, Node) else x)
 
 
 @compatibility(is_backward_compatible=True)
-def map_aggregate(a: Argument, fn: Callable[[Argument], Argument]) -> Argument:
+def map_aggregate(a: ArgumentT, fn: Callable[[Argument], Argument]) -> ArgumentT:
     """
-    Apply fn to each Node appearing arg. arg may be a list, tuple, slice, or dict with string keys.
+    Apply fn recursively to each object appearing in arg.
+
+    arg may be a list, tuple, slice, or dict: the return value will have
+    the same type and structure.
     """
+    result: Argument
+
     if isinstance(a, tuple):
-        t = tuple([map_aggregate(elem, fn) for elem in a])
-        # Support NamedTuple (if it has `_fields`) by repacking into original type.
-        return t if not hasattr(a, "_fields") else type(a)(*t)  # type: ignore[arg-type]
+        it = (map_aggregate(elem, fn) for elem in a)
+        result = type(a)(*it) if hasattr(a, "_fields") else tuple(it)
     elif isinstance(a, list):
-        return immutable_list([map_aggregate(elem, fn) for elem in a])
+        result = immutable_list(map_aggregate(elem, fn) for elem in a)
```

Comment:
torch compile might need this to be a listcomp

Reply:
@Skylion007 Indeed, this was a problem, and so was the dict comprehension a few lines below. I shouldn't have monkeyed with it.

There is work being done right now with dynamo to make it better handle generators of all types, so perhaps one day I'll come back to these for a lark and modernize them once it works.